xformers/components/input_projection.py

```diff
@@ -0,0 +1,206 @@
+            )
+        else:
+            # The dimensions are different, use seperate buffers
+            self.q_proj = nn.Linear(
+                query_proj_params.in_features,
+                self.out_features,
+                bias=query_proj_params.bias,
+            )
+            self.k_proj = nn.Linear(
+                key_proj_params.in_features,
+                self.out_features,
+                bias=key_proj_params.bias,
+            )
+            self.v_proj = nn.Linear(
+                value_proj_params.in_features,
+                self.out_features,
+                bias=value_proj_params.bias,
+            )
+
+        # - multi-head attention specific init for the weights and biases
+        self._reset_parameters()
+
+    def _reset_parameters(self):
+        if self.self_attention:
+            # Init each of the projection blocks independently
+            out_features = self.in_proj.weight.shape[0] // 3
+
+            with torch.no_grad():
+                self.in_proj.weight[:out_features, :] = self._init_weights(
+                    self.q_p_params, self.in_proj.weight[:out_features, :]
+                )
+                self.in_proj.weight[
+                    out_features : 2 * out_features, :
+                ] = self._init_weights(
+                    self.q_p_params,
+                    self.in_proj.weight[out_features : 2 * out_features, :],
+                )
+                self.in_proj.weight[2 * out_features :, :] = self._init_weights(
+                    self.q_p_params, self.in_proj.weight[2 * out_features :, :]
+                )
+        else:
+            self.q_proj.weight = self._init_weights(self.q_p_params, self.q_proj.weight)
+            self.k_proj.weight = self._init_weights(self.k_p_params, self.k_proj.weight)
+            self.v_proj.weight = self._init_weights(self.v_p_params, self.v_proj.weight)
+
+    @staticmethod
+    def _init_weights(params: InProjParams, weights: torch.Tensor, gain: float = 1.0):
+        if params.small_init:
+            return small_init_(weights, gain=gain)
+        else:
+            return xavier_uniform_(weights, gain=gain)
+
+    def forward(
+        self,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        if self.self_attention:
+            assert self.in_proj is not None
+            assert id(query) == id(key) and id(query) == id(
+                value
+            ), "Self attention was requested, but query, key and value differ"
+
+            # All the inputs have the same dimension, a small optimization is to compute the projections all in one go
+            # We use a small trick here, to make sure that the output buffer -which we'll split-
+            # produces contiguous buffers from the get go
+            B, S, _ = query.shape
+            qkv = self.in_proj.weight @ query.movedim(-1, 0).flatten(1, 2)
```

Comment:
@dianaml0 this is the trick that I was thinking about, it turns out that you don't need a custom tensor for that, just be a little careful with the dimensions and transposes. This plays well with the perf, incoming numbers :) this removes a read and a write, looks like nothing but I'm glad

Reply:
(the key here is that the .contiguous() is gone, which removes the need for a read and transpose for each of the q,k,v buffers. This is really free lunch AFAIK)

Reply:
hmm, somehow this seems to break autograd :(